src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    currentState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const prop of action.keysToRemove) {
+        delete currentState[prop];
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    states.push(currentState);
```

Comment:
According to the checklist, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the switch block, after the current action is processed. This ensures that the state is correctly captured after all actions are applied.